src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -108,35 +121,176 @@ def __init__(
         self,
         commit: Callable[[Mapping[Partition, Position]], None],
         max_batch_size: int,
-        max_batch_time: int,
     ) -> None:
-        self.__commit = commit
-        self.__max_batch_size = max_batch_size
-        self.__max_batch_time = timedelta(milliseconds=max_batch_time)
-        self.__messages_since_last_commit = 0
-        self.__last_commit = datetime.now()
-        self.__ready_to_commit: MutableMapping[Partition, Position] = {}
-        self.__closed = False
+        self._closed: bool = False
+        self._commit = commit
+        # XXX(iker): these max* variables currently don't behave as maximums,
+        # but rather minimums. We should fix this at some point.
+        self._max_batch_size: int = max_batch_size
+
+        self._billing_topic = Topic(settings.KAFKA_OUTCOMES_BILLING)
+        self._billing_producer = self._get_billing_producer()
+        # Especially on incidents generating big backlogs, we must expect a lot
+        # of removals from the beginning of a big FIFO. A dequeue provides O(1)
+        # time on removing items from the beginning; while a regular list takes
+        # O(n).
+        self._ongoing_billing_outcomes: Deque[MetricsBucketBilling] = deque()
+        self._ready_to_commit: MutableMapping[Partition, Position] = {}
+        self._messages_ready_since_last_commit: int = 0
+
+    def _get_billing_producer(self) -> KafkaProducer:
+        servers = _get_bootstrap_servers(topic=self._billing_topic.name, force_cluster=None)
+        return KafkaProducer(
+            build_kafka_configuration(default_config={}, bootstrap_servers=servers)
+        )
 
     def poll(self) -> None:
         if self._should_commit():
             self._bulk_commit()
 
+    def _should_commit(self) -> bool:
+        self._mark_commit_ready()
+
+        if not self._ready_to_commit:
+            return False
+        if self._messages_ready_since_last_commit >= self._max_batch_size:
+            return True
+        return False
+
+    def _mark_commit_ready(self) -> None:
+        """Removes completed futures at the beginning of
+        ongoing_billing_outcomes and adds them to ready_to_commit.
+
+        If a future has errored, log an error but add to ready_to_commit
+        anyways.
+        """
+        while self._ongoing_billing_outcomes:
+            produce_billing, metrics_msg = self._ongoing_billing_outcomes[0]
+
+            if produce_billing:
+                if not produce_billing.done():
+                    break
+                # XXX(iker): in tests .exception() may return a mock, which
+                # resolves to a truthy value. Explicitly checking if it's an
+                # exception removes this log from tests. We may want to improve
+                # mocking in the tests and only check for a None value here.
```

Comment:
Not sure I follow. Can you mock the exception() function to return whatever you need it to. Ideally I think that needing to write you code in a very specific way just to make the tests pass should be avoided.

Reply:
I agree with your comment. This is the bad, lazy and fast way of solving it, which I recognize is not ideal. `.exception()` is called on a future, produced by a mocked object, returned by the function mocked in the test; so I decided to not do the whole mocking dance and take this shortcut instead to solve it faster.